### ISSUE TITLE
StatefulSet's PersistentVolumeClaims have incorrect labels

### ISSUE THREAD

**Issue by llamasoft** [kubernetes/kubernetes#74916]
**What happened**:  
When creating a StatefulSet that uses volumeClaimTemplates, the resulting PersistentVolumeClaims have incorrect labels.  Instead of using the labels defined in their metadata, they use the labels from the StatefulSet's selector.

**What you expected to happen**:
I expect the created PersistentVolumeClaims to have the labels defined in their metadata.

**How to reproduce it (as minimally and precisely as possible)**:

```
apiVersion: v1
kind: Service
metadata:
  name: bugdemo-service
  labels:
    app: bugdemo-service
    extraLabel: fromService
spec:
  ports:
    - name: nginx
      targetPort: 80
      port: 80
  clusterIP: None
  selector:
    app: bugdemo-statefulset
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: bugdemo-statefulset
  labels:
    app: bugdemo-statefulset
    extraLabel: fromStatefulSet
spec:
  selector:
    matchLabels:
      app: bugdemo-pod
  serviceName: bugdemo-service
  replicas: 1
  template:
    metadata:
      labels:
        app: bugdemo-pod
        extraLabel: fromPodTemplate
    spec:
      containers:
      - name: bugdemo-nginx
        image: nginx:stable
        resources:
          requests:
            cpu: "100m"
            memory: "1Gi"
        ports:
          - containerPort: 80
        volumeMounts:
          - name: nginx-data
            mountPath: /var/log/dummy
  volumeClaimTemplates:
    - metadata:
        name: nginx-data
        # In the resulting PersistentVolumeClaim, these
        # labels are replaced with the StatefulSet's selector
        labels:
          app: bugdemo-volumeclaim
          extraLabel: fromVolumeClaimTemplate
      spec:
        accessModes:
          - ReadWriteOnce
        storageClassName: standard
        resources:
          requests:
            storage: "1Gi"
```

Resulting PersistentVolumeClaim:

```
$ kubectl get pvc --show-labels
NAME                               STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE   LABELS
nginx-data-bugdemo-statefulset-0   Bound    pvc-7917ee4f-3ec7-11e9-917f-42010aXXXXXX   1Gi        RWO            standard       1m    app=bugdemo-pod
```

**Environment**:
- Kubernetes version (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.2", GitCommit:"cff46ab41ff0bb44d8584413b598ad8360ec1def", GitTreeState:"clean", BuildDate:"2019-01-13T23:15:13Z", GoVersion:"go1.11.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"11+", GitVersion:"v1.11.7-gke.4", GitCommit:"618716cbb236fb7ca9cabd822b5947e298ad09f7", GitTreeState:"clean", BuildDate:"2019-02-05T19:22:29Z", GoVersion:"go1.10.7b4", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration: GKE
- Install tools: `kubectl`
- Others:

@kubernetes/sig-apps-apps . 
/sig apps

**Comment by llamasoft** [comment]
@kubernetes/sig-contributor-experience-apps

**Comment by cppforlife** [comment]
+1 just ran into this myself.

**Comment by kow3ns** [comment]
As in #74941. Even thought the StatefulSet controller looks for the PVC by name (not by label) we generate labels for the PVC based on the StatefulSet's selector. Removing this would be a backward incompatible change to a v1 API contract (can't do that). Adding additional labels might be acceptable, but, If you want to change the behavior of the controller wrt to this, I'd like to see a (small) KEP for it so it can be discussed a bit more broadly in the SIG.

**Comment by llamasoft** [comment]
I very much think this would be a good discussion to have.  If you could, please let me know what's required to start a KEP on this topic.

As I see it, the current implementation is in violation of the API: the [StatefulSet's](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#statefulsetspec-v1-apps) volumeClaimTemplates data type is listed as "[PersistentVolumeClaim](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#persistentvolumeclaim-v1-core) array".  A PersistentVolumeClaim's metadata is listed as a standard [ObjectMeta](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#objectmeta-v1-meta) type.  
The fact that the PersistentVolumeClaim's metadata is being handled differently when defined in a StatefulSet means that the PersistentVolumeClaim API isn't being properly followed in that context.

I'd also like to argue that the change wouldn't be backwards incompatible.  The StatefulSet's volumeClaimTemplates description even says that the matching is done on name: "Every claim in this list must have at least one matching (by name) volumeMount in one container in the template."  
Respecting the PersistentVolumeClaim's metadata wouldn't necessarily cause any incompatibility from a technical perspective (although it may affect some users' _implementations_ that rely on it).

In summary, the current implementation violates the principle of least surprise, makes it impossible to label individual volume claims separately, and it directly contradicts the API specification.

**Comment by kow3ns** [comment]
@llamasoft The KEP process is described [here](https://github.com/kubernetes/enhancements/). You can open a KEP via a PR against the SIG Apps repository. You can use me as your approver. 

As to the compatibility of the current PR, the current label generation is v1 and has worked the same way for several years now. Users that select the PVCs created by the controller have a reasonable expectation that this should continue to function in subsequent releases. If you add additional labels but generate labels that match the selector this would allow for additional labels while maintaining compatibility (imo).
